Much better collection permission performance

johnswanson opened the pull request:
## Migrations: add and populate indexed columns `perm_value`, `perm_type`, and `collection_id` to `permissions`

These fields allows us to efficiently run queries based on collection permissions in the DB without string manipulation. Keeping the table as `permissions` allows us to do this migration in-place. **Note that in some cases collections may have been deleted from the database without deleting the associated permissions row (since there was no foreign key before).** We need to be defensive here: if we have a permissions row without a corresponding collection, delete the row before running the rest of the migration.

## Write `perm_value`, `perm_type`, and `collection_id` for new collection permissions

A very simple `before-insert` method sets these fields before a collection permission is written to the DB.

## Replace `collection/permissions-set->visible-collection-ids` and `collection/visible-collection-ids->honeysql-filter-clause` with `collection/honeysql-filter-clause`

Previously, just about everywhere we used `permissions-set->visible-collection-ids`, what we were essentially doing was an in-app join: select all the collection IDs you have permission on, then convert it to a SQL clause like `WHERE collection_id IN ( all of those collection IDs)`.

Replace both of these with `honeysql-filter-clause`, which uses the new fields we added to `permissions` above to construct a honeysql filter representing "all the collections I have permissions on", without needing to round-trip them to the application and back to the DB.

Of course, we can then write a function `visible-collection-ids`, which uses `honeysql-filter-clause`, for those cases where we *do* actually need the whole bunch in the application (we use this, for example, when constructing the `effective-location` for a collection).

I also added one more toggle to the `VisibilityConfig` that's passed into the `honeysql-filter-clause` (and used to be passed to `permissions-set->visible-collection-ids`), allowing you to select only the effective children of some collection.

The following graph shows the old vs new code doing a search for a non-admin:

```
(search/search
                                          (search/search-context
                                           {:archived                            false
                                            :created-at                          nil
                                            :created-by                          nil
                                            :current-user-id                     api/*current-user-id*
                                            :current-user-perms                  @api/*current-user-permissions-set*
                                            :filter-items-in-personal-collection nil
                                            :last-edited-at                      nil
                                            :last-edited-by                      nil
                                            :limit                               1
                                            :model-ancestors?                    false
                                            :models                              #{"dashboard" "dataset" "card"}
                                            :offset                              nil
                                            :search-native-query                 nil
                                            :search-string                       nil
                                            :table-db-id                         nil
                                            :verified                            nil
                                            :ids                                 nil}))
```

X-axis is number of collections, Y-axis is ms to complete. Also note that the timing jump for the old code coincided with it beginning to fail with an error about too many SQL parameters.

<img width="909" alt="Screenshot 2024-08-21 at 12 52 17" src="https://github.com/user-attachments/assets/442ded6f-bcee-42c1-b0bd-a1c65be0a2b4">

This is the fix for https://github.com/metabase/metabase/issues/46510 (but needs to be backported, as that issue is for v49)

johnswanson committed:
pvt `collection-ids->honeysql-filter-clause`

Previously `collection` had two public functions:

`collection/permissions-set->visible-collection-ids` took a permissions
set and returned a set of collection IDs you should be allowed to see.

`collection/visible-collection-ids->honeysql-filter-clause` took the
output of the above and returned a honeysql filter clause that would
restrict `collection_id` (or another field you passed in) to be in the
set.

But the caller shouldn't need to know how exactly we're constructing the
honeysql filter clause. They just want the filter clause - deciding how
exactly to build it (get the set of collection IDs I can view, then
filter on it) is an implementation detail that we might move away from.
For example, hopefully we'll move to a permissions model where we can
essentially let the database figure out the set of IDs you have access
to, and the filter clause can use a subselect.

johnswanson committed:
Some small perf improvements for coll perms

This is a minor performance improvement but probably worthwhile.

There are a few changes:

- add a new knob to `permissions-set->visible-collection-ids`. If an
`effective-child-of` is passed as part of the config map, it will only
return collection IDs that are an effective child of the passed parent.
Previously, we would find the complete set of visible collection IDs,
and then exclude things that weren't an effective child, which basically
involved passing *every* visible collection ID to the query *twice*:
once as `collection_id IN [visible-collection-ids]` and once as
`location NOT LIKE [a collection ID]` for a large subset of the visible
collection IDs. Now, when looking for effective children, we just find
the set of collection IDs that are effective children of the parent, and
only need to pass it off to SQL once.

- when turning a set of visible collection IDs to a honeysql filter
clause, check to see which is bigger: the set of visible collection IDs,
or the set of *invisible* collection IDs. Use the smaller set to create
the filter clause. This way, if you have access to 99.9% of collections,
we'll just pass the *remaining* .1% of collections to the query, like
`collection_id NOT IN [invisible-collection-ids]` instead of
`collection_id IN [visible-collection_ids]`.

Testing this out locally with 5000 collections, endpoints that
previously timed out with an error (too many SQL params) took ~1.5s to
return results. Still not ideal, but better.

johnswanson committed:
Provide alternative to `/search` for `has_model`

Add `has_model` to `/api/user/current` to allow us to avoid using the
`/search` endpoint to see whether the user has a readable model.

Using `has-readable-model?`, which joins directly to `permissions`,
rather than `collection/honeysql-filter-clause`, works much better
since:

- we're just trying to see whether something with the given
conditions *exists*, and

- the number of visible collections may be quite large

Testing with 15,000 readable collections, and a single dashboard / card
buried inside one of them, this took ~21-45ms locally. The current
version takes ~2s to populate only `has_question_and_dashboard`.

johnswanson committed:
improve performance of `honeysql-filter-clause`

Write a subselect that finds the set of collection IDs you have access
to, rather than selecting them in code and then passing them back to the database.

johnswanson committed:
fix honeysql-filter-clause

johnswanson committed:
partial collection permissions migration

Rather than migrating collection `permissions` to a new
`collection_permissions` table, we can do an in-place migration. A
simple `before-insert` method sets `perm_type`, `perm_value`, and
`collection_id` correctly.

Note that this does not currently handle *root* collection permissions
of any kind. There are really 4 possible cases there, the permission
`object` is:

- `/collection/root/`,

- `/collection/root/read/`,

- `/collection/namespace/snippets/root/`, or

- `/collection/namespace/snippets/root/read/`.

When we do migrate these over, we'll need to create a `namespace`
column. Presumably we'll set `collection_id` to `NULL` but the
`namespace` will allow us to distinguish between having permissions on
the snippets root and the regular collection root.

For now, we can just migrate over ordinary collection permissions.

johnswanson committed:
Use `collection/honeysql-filter-clause` more

One of my earlier performance improvements was to roll my own version of
this to avoid getting all collection IDs in memory. Now that we're doing
that in `collection/honeysql-filter-clause` as well, there's no need!

johnswanson committed:
Lint migrations

johnswanson committed:
Fix delete cascade for `permissions.collection_id`

johnswanson committed:
Don't delete indexes, let them go with the column

johnswanson committed:
Rip out hand-rolled collection perms check

From what I can tell, this is doing exactly the same thing as
`collection/honeysql-filter-clause`, just less efficiently.

johnswanson committed:
WIP: remove `collection-id->collection` entirely

This was a misstep (on my part). I added it thinking that the number of
collections should always be small enough to fit in memory and just sort
in the app. But in my benchmarking I've found that this performs really
poorly for large numbers of collections - which makes sense. It's really
not necessary - we can do this kind of sorting in the DB, which is more
reasonable anyway.

johnswanson committed:
cljfmt

johnswanson committed:
test fix

johnswanson committed:
More test fixes

johnswanson committed:
linty mclinterson

johnswanson committed:
Prep for CR

Add some more docstrings, clean up existing docstrings, and comment the
crazy bits

noahmoss commented on src/metabase/api/user.clj:
```diff
@@ -299,16 +299,13 @@
-                         (collection/permissions-set->visible-collection-ids @api/*current-user-permissions-set*))
-        entity-exists? (fn [model] (t2/exists? model
-                                               {:where [:and
-                                                        [:= :archived false]
-                                                        coll-ids-filter
-                                                        (mi/exclude-internal-content-hsql model)]}))]
```
Looks like you might have lost the `:= :archived false` and `exclude-internal-content-hsql` clauses here?

noahmoss commented on src/metabase/api/collection.clj:
```diff
@@ -641,14 +633,10 @@
   [parent-coll colls]
-  (let [visible-collection-ids (collection/permissions-set->visible-collection-ids @api/*current-user-permissions-set*
-                                                                                   {:include-archived-items :all
-                                                                                    :archive-operation-id nil
-                                                                                    :permission-level :read})
```
Never mind — I see that we now have default config values.

noahmoss commented on src/metabase/models/collection.clj:
```diff
@@ -311,161 +480,173 @@
+   ;; we're not looking for the children of a collection (root definitely isn't a child!)
+   (not (:effective-child-of visibility-config))))
+
+(mu/defn honeysql-filter-clause
```
nit - I think the naming could be a little more specific. Maybe `collection-filter-clause` or `collection-permissions-filter-clause`? It's not a huge deal but `honeysql-filter-clause` could be referring to anything

noahmoss commented on src/metabase/models/permissions.clj:
```diff
@@ -345,9 +345,20 @@
 
 (derive :model/Permissions :metabase/model)
 
+(defn- maybe-break-out-permission-data
```
Really like this migration approach!

noahmoss reviewed (changes requested):
Could you add a migration test (in `metabase.db.schema-migrations-test`) that asserts that the new `permissions` fields are populated correctly for existing permissions?

Otherwise, changes look good! Called out a few spots that looked a little suspicious to me but I might be misreading them.

johnswanson committed:
CR response

- rename `honeysql-filter-clause` to `visible-collection-filter-clause`

- fix the `has_question_and_dashboard` key on users, which I broke

johnswanson committed:
Delete perms for nonexistent collections

Also, add a test of the behavior of the new permissions migration.
